README
DOSBox v0.74-2 Manual (always use the latest version from www.dosbox.com)



=====
NOTE:
=====

While we are hoping that one day DOSBox will run all programs ever made for
the PC, we are not there yet.
At present, DOSBox running on a high-end machine will roughly be the equivalent
of a Pentium I PC. DOSBox can be configured to run a wide range of DOS games,
from CGA/Tandy/PCjr classics up to games from the Quake era.



======
INDEX:
======

1. Quickstart
2. Start (FAQ)
3. Command Line Parameters
4. Internal Programs
5. Special Keys
6. Joystick/Gamepad
7. KeyMapper
8. Keyboard Layout
9. Serial Multiplayer feature
10. How to speed up/slow down DOSBox
11. Troubleshooting
12. DOSBox Status Window
13. The configuration (options) file
14. The language file
15. Building your own version of DOSBox
16. Special thanks
17. Contact



==============
1. Quickstart:
==============

Type INTRO in DOSBox for a quick tour.
It is essential that you get familiar with the idea of mounting, DOSBox does not
automatically make any drive (or a part of it) accessible to the emulation. See
the FAQ entry "How to start?" as well as the description of the MOUNT command
(Section 4: "Internal Programs"). If you have your game on a CD-ROM you may try
this guide: https://www.vogons.org/viewtopic.php?t=8933



===============
2. Start (FAQ):
===============

START:      How to start?
AUTOMATION: Do I always have to type these "mount" commands?
FULLSCREEN: How do I change to fullscreen?
FULLSCREEN: My fullscreen is too large.
CD-ROM:     My CD-ROM doesn't work.
CD-ROM:     The game/application can't find its CD-ROM.
MOUSE:      The mouse doesn't work.
SOUND:      There is no sound.
SOUND:      What sound hardware does DOSBox presently emulate?
SOUND:      The sound stutters or sounds stretched/weird.
KEYBOARD:   I can't type \ or : in DOSBox.
KEYBOARD:   Right Shift and "\" doesn't work in DOSBox. (Windows only)
KEYBOARD:   The keyboard lags.
CONTROL:    The character/cursor/mouse pointer always moves into one direction!
SPEED:      The game/application runs much too slow/too fast!
CRASH:      The game/application does not run at all/crashes!
CRASH:      DOSBox crashes on startup!
GAME:       My Build game(Duke3D/Blood/Shadow Warrior) has problems.
SAFETY:     Can DOSBox harm my computer?
OPTIONS:    I would like to change DOSBox's options.
HELP:       Great Manual, but I still don't get it.



START: How to start?
    At the beginning you've got a Z:\> instead of a C:\> at the prompt.
    You have to make your directories available as drives in DOSBox by using
    the "mount" command. For example, in Windows "mount C D:\GAMES" will give
    you a C drive in DOSBox which points to your Windows D:\GAMES directory
    (that was created before). In Linux, "mount c /home/username" will give you
    a C drive in DOSBox which points to /home/username in Linux.
    To change to the drive mounted like above, type "C:". If everything went
    fine, DOSBox will display the prompt "C:\>".


AUTOMATION: Do I always have to type these commands?
    In the DOSBox configuration file is an [autoexec] section. The commands
    present there are run when DOSBox starts, so you can use this section
    for the mounting. Look at Section 13: "The configuration (options) file".


FULLSCREEN: How do I change to fullscreen?
    Press alt-enter. Alternatively: Edit the configuration file of DOSBox and
    change the option fullscreen=false to fullscreen=true. If fullscreen looks
    wrong in your opinion: Play with the options: fullresolution, output and
    aspect in the configuration file of DOSBox. To get back from fullscreen
    mode: Press alt-enter again.


FULLSCREEN: My fullscreen is too large.
    This is can be a problem on Windows 10, if you have display scaling
    set to a value above 100%. Windows in that case will resize the screen
    on top of dosbox resizing the screen, which can happen for the output:
    ddraw, opengl, openglnb, overlay. You can disable this Windows behaviour
    by enabling a specific compatibility setting:
 
    - Right-click the DOSBox icon and select "Properties".
    - Go to the "Compatibility" tab.
    - Click on "Change high DPI settings".
    - Tick "Override high DPI scaling behaviour" and set it to "Application".
    - Apply the changes by clicking on "OK".

    Unfortunately, this compatibility option causes some side effects in
    windowed mode, and in this case you will need to change the resolution
    in the config/Options file for windowresolution (e.g. 1024x768).

    Alternatively, you can disable the display scaling and or use a lower
    fullresolution value.

CD-ROM: My CD-ROM doesn't work.
    To mount your CD-ROM in DOSBox you have to specify some additional options
    when mounting the CD-ROM.
    To enable CD-ROM support (includes MSCDEX) in Windows:
      - mount d f:\ -t cdrom
    in Linux:
      - mount d /path/to/cdrom -t cdrom

    explanation: - d   driveletter you will get in DOSBox (d is the best,
                          don't change it!)
                 - f:\ location of CD-ROM on your PC. In most cases it will
                          be d:\ or e:\

    See also the next question: The game/application can't find its CD-ROM.


CD-ROM: The game/application can't find its CD-ROM.
    Be sure to mount the CD-ROM with -t cdrom switch, this will enable the
    MSCDEX interface required by DOS games to interface with CD-ROMs.
    Also try adding the correct label (-label LABEL) to the mount command,
    where LABEL is the CD-label (volume ID) of the CD-ROM.

    Try creating a CD-ROM image (preferably CUE/BIN pair) and use the
    DOSBox's internal IMGMOUNT tool to mount the image (the CUE sheet).
    This enables very good low-level CD-ROM support on any operating system.


MOUSE: The mouse doesn't work.
    Usually, DOSBox detects when a game uses mouse control. When you click on
    the screen it should get locked (confined to the DOSBox window) and work.
    With certain games, the DOSBox mouse detection doesn't work. In that case
    you will have to lock the mouse manually by pressing CTRL-F10.


SOUND: There is no sound.
    Be sure that the sound is correctly configured in the game. This might be
    done during the installation or with a setup/setsound utility that
    accompanies the game. First see if an autodetection option is provided. If
    there is none try selecting SoundBlaster or SoundBlaster 16 with the default
    settings being "address=220 irq=7 dma=1" (sometimes highdma=5). You might
    also want to select Sound Canvas/SCC/MPU-401/General MIDI/Wave Blaster
    at "address=330 IRQ=2" as music device.
    The parameters of the emulated sound cards can be changed in the DOSBox
    configuration file.
    If you still don't get any sound set the core to normal in DOSBox
    configuration and use some lower fixed cycles value (like cycles=2000). Also
    assure that your host operating sound does provide sound.
    In certain cases it might be useful to use a different emulated sound device
    like a SoundBlaster Pro (sbtype=sbpro1 in the DOSBox configuration file) or
    the Gravis Ultrasound (gus=true).


SOUND: What sound hardware does DOSBox presently emulate?
    DOSBox emulates several legacy sound devices:
    - Internal PC speaker/Buzzer
      This emulation includes both the tone generator and several forms of
      digital sound output through the internal speaker.
    - Creative CMS/Gameblaster
      The is the first card released by Creative Labs(R).  The default
      configuration places it on address 220. It is disabled by default.
    - Tandy 3 voice
      The emulation of this sound hardware is complete with the exception of
      the noise channel. The noise channel is not very well documented and as
      such is only a best guess as to the sound's accuracy. It is disabled as
      default.
    - Tandy DAC
      Some games may require turning off SoundBlaster emulation (sbtype=none)
      for better Tandy DAC sound support. Don't forget to set the sbtype back to
      sb16 if you don't use Tandy sound.
    - Adlib
      This emulation is almost perfect and includes the Adlib's ability to
      almost play digitized sound. Placed at address 220 (also on 388).
    - SoundBlaster 16 / SoundBlaster Pro I & II / SoundBlaster I & II
      By default DOSBox provides SoundBlaster 16 level 16-bit stereo sound.
      You can select a different SoundBlaster version in the configuration of
      DOSBox. AWE32 music is not emulated as you can use MPU-401 instead
      (see below).
    - Disney Sound Source and Covox Speech Thing
      Using the printer port, this sound device outputs digital sound only.
      Placed at LPT1
    - Gravis Ultrasound
      The emulation of this hardware is nearly complete, though the MIDI
      capabilities have been left out, since an MPU-401 has been emulated
      in other code. For Gravis music you also have to install Gravis drivers
      inside DOSBox. It is disabled by default.
    - MPU-401
      A MIDI passthrough interface is also emulated. This method of sound
      output will only work when used with external device/emulator.
      Every Windows XP/Vista/7 and Mac OS X has got a default emulator 
      compatible with: Sound Canvas/SCC/General Standard/General MIDI/Wave 
      Blaster. A different device/emulator is needed for 
      Roland LAPC/CM-32L/MT-32 compatibility.


SOUND: The sound stutters or sounds stretched/weird.
    You may be using too much CPU power to keep DOSBox running at the current
    speed. You can lower the cycles, skip frames, reduce the sampling rate of
    the respective sound device, increase the prebuffer. See Section 13: "The
    configuration (options) file".
    If you are using 'cycles=max' or 'cycles=auto', then make sure that there is
    no background processes interfering! (especially if they access the harddisk)
    Also look at Section 10: "How to speed up/slow down DOSBox" as well as
    Section 11: "Troubleshooting".


KEYBOARD: I can't type \ or : in DOSBox.
    This can happen in various cases, like your host keyboard layout does not
    have a matching DOS layout representation (or it was not correctly
    detected), or the key mapping is wrong.
    Some possible fixes:
      1. Use / instead, or ALT-58 for : and ALT-92 for \.
      2. Change the DOS keyboard layout (see Section 8: "Keyboard Layout").
      3. Add the commands you want to execute to the [autoexec] section
         of the DOSBox configuration file.
      4. Open the DOSBox configuration file and change the usescancodes entry.
      5. Switch the keyboard layout of your operating system.

    Note that if the host layout can not be identified, or keyboardlayout is
    set to none in the DOSBox configuration file, the standard US layout is
    used. In this configuration try the keys around "enter" for the key \
    (backslash), and for the key : (colon) use shift and the keys between
    "enter" and "L".


KEYBOARD: Right Shift and "\" doesn't work in DOSBox. (Windows only)
    This may happen if Windows thinks that you have more than one keyboard
    connected to your PC when you use some remote control devices.
    To verity this problem run cmd.exe, navigate to DOSBox program folder
    and type:
    set sdl_videodriver=windib
    dosbox.exe
    check whether keyboard started to work properly. As windib is slower it is
    best to use one of the two solutions provided here:
    https://www.vogons.org/viewtopic.php?t=24072


KEYBOARD: The keyboard lags.
    Lower the priority setting in the DOSBox configuration file, for example
    set "priority=normal,normal". You might also want to try lowering the
    cycles (use a fixed cycle amount to start with, like cycles=10000).


CONTROL: The character/cursor/mouse pointer always moves into one direction!
    See if it still happens if you disable the joystick emulation,
    set joysticktype=none in the [joystick] section of your DOSBox
    configuration file. Maybe also try unplugging any joystick/gamepad.
    If you want to use the joystick in the game, try setting timed=false
    and be sure to calibrate the joystick (both in your OS as well as
    in the game or the game's setup program).


SPEED: The game/application runs much too slow/too fast!
    Look at Section 10: "How to speed up/slow down DOSBox" for more
    information.


CRASH: The game/application does not run at all/crashes!
    Look at Section 11: "Troubleshooting".


CRASH: DOSBox crashes on startup!
    Look at Section 11: "Troubleshooting".


GAME: My Build game(Duke3D/Blood/Shadow Warrior) has problems.
    First of all, try to find a port of the game. Those will offer a better
    experience. To fix the graphics problem that occurs in DOSBox on higher
    resolutions: Open the configuration file of DOSBox and search for
    machine=svga_s3. Change svga_s3 to vesa_nolfb
    Change memsize=16 to memsize=63


SAFETY: Can DOSBox harm my computer?
    DOSBox can not harm your computer more than any other resource demanding
    program. Increasing the cycles does not overclock your real CPU.
    Setting the cycles too high has a negative performance effect on the
    software running inside DOSBox.


OPTIONS: I would like to change DOSBox's options.
    Look at Section 13: "The configuration (options) file".


HELP: Great Manual, but I still don't get it.
    For more questions read the rest of this Manual. You may also look at:
    guides located at https://www.vogons.org/viewforum.php?f=53
    the wiki of DOSBox https://www.dosbox.com/wiki/
    the site/forum: https://www.dosbox.com



===========================
3. Command Line Parameters:
===========================

An overview of the command line options you can give to DOSBox. Although
in most cases it is easier to use DOSBox's configuration file instead.

NEWS
0.74-3
A security release for DOSBox 0.74:
- Fixed that a very long line inside a bat file would overflow the parsing 
  buffer (CVE-2019-7165 by Alexandre Bartel)
- Added a basic permission system so that a program running inside DOSBox 
  can't access the contents of /proc (e.g. /proc/self/mem) when / or /proc
  were (to be) mounted (CVE-2019-12594 by Alexandre Bartel)  
- Several other fixes for out of bounds access and buffer overflows.
- Some fixes to the OpenGL rendering.

It's recommended to use config -securemode when dealing with untrusted
files. 
See the Changelog for all the changes.

0.74-2
A maintenance release for DOSBox 0.74, which solves the following problems:

- Windows: Fix auto/max cycles algorithm on Windows 7, which helps with 
           stuttering audio.
- Mac OS X: Bring a 64 bit version and improve performance.
- Linux: Fix the 64bit dynrec cpu core and a lot of compilation problems. 
         Add patches for the WINE Team.

The game compatibility should be identical to 0.74.
See the Changelog for a full list of changes.

0.74
  - Several small game specific fixes/hacks/support. (Offensive,
    Roadhog, GTA installer, Kingdom O' Magic soundcard detection, 
    Pirate booter, Armored Fist installer)
  - Add the S3-specific 640x480 256 color mode. (fixes regression in "Wooden 
    Ships and Iron Men" and "I Have No Mouth  And I Must Scream")
  - Fix a stack overflow that could crash DOSBox.
  - Add fake microphone input. (fixes Talking Parrot)
  - Modify adlib turn off code, so that it doesn't turn off in 
    cases where the same sound is repeated over and over again.
  - Several small fixes to the CDROM audio code. (HOMM2, Redneck 
    Rampage and others)
  - Several improvements to the CDROM emulation code. (fixes Alpha
    Storm and  GT Racing 97)
  - Some small CPU fixes that might fix something.
  - Handle opcode 0xff subcode 7 as invalid instruction. (fixes dif-2 & others)
  - Some hercules fixes. (Testdrive)
  - Improve support for blanked parts that wrap around to the start of
    the screen. (fixes Magic Circle demo and Sid&Al)
  - Remove old OPL cores as the new ones seem to work very nicely.
  - Modify movie recording code so that the movies aren't corrupt when
    you exit dosbox without stopping the movie.
  - Change RGB3x scaler to look more pretty.
  - Improve initial register values compatility of the GUS.
  - Added autodetection for Gameblaster. (games can find it now)
  - Change render preferences a bit to be more compatible with Windows 7.
  - Add DOS fixes to terminate program. (fixes Fortune Teller) 
  - Add FFREEP. (fixes Trucks)
  - Improve FPU ST80 in C mode when writing zero. (fixes Antigok)
  - Add special int10 scanline function. (fixes mz700 and probably lots
    of games that mess with them)
  - Fix scrolling in rarely used video modes. (fixes Orphee)
  - Modify game specific hacks a bit so that Kick off 3 works again.
  - Lots of fixes to the INT10 video parameter table. (Seven spirits 
    of ra and others)
  - Add VGA blanking in machine=vgaonly. (used by Alien Carnage)
  - CGA, PCJr, Tandy: Add video blanking, change display start latch 
    timing, sync pulse width correction.
  - PCJr, Tandy: implement vertical retrace interrupt.
  - PCJr, CGA: line-by-line video emulation.
  - PCJr: support on-screen change of color modes 4medium to 16low. (used
    by Ghostbusters booter)
  - Hercules: Add green and amber monochrome support.
  - All machines: only update the video timing when needed. (Jungle Hunt,
    others that synchronize to the video screen might profit)
  - Several small DOS fixes.
  - Some UMB related fixes. (The Legacy without UMB)
  - Fix version number of DSP for SB 1.5. (fixes a few games)
  - Several VGA emulation improvements. (Allertone football manager)
  - Some Tandy fixes. (Mech Warrior)
  - Small improvements and fixes to the OPL emulation.
  - Add low level Tandy DAC emulation.
  - Some EMS fixes. (fixes Mortal Kombat and others)
  - Change SoundBlaster DSP reset mechanism, add sb irq acknowledge logic.
    (fixes stmik-based applications)
  - Some interrupt pointer location modifications. (fixes Tinker Tales)
  - Some fixes to the BOOT code. (fixes Last Mission)
  - Respect write-only file information. (fixes Champions of Zulala)
  - Some RTC fix. (fixes Tully Bodine and others)
  - Improve mouse emulation to work better with Water World.
  - Hopefully fix the translation of the configuration file.
  - Speed up and fixes for the recompiler core. (pitfall2 pcjr)
  - Change memory start location. (fixes 7th Guest installer)
  - Several fixes to the batch file handling. (Shift and 
    use the typed first %0 instead of the parsed %0)
  - Improve file redirection and redirected line ends. (fixes 
    Phantasmagoria 2 DOS installer)
  - Fix compilation with new MAC os X version.
  - Add 16C550A FIFO support to the serial port emulation.
  - Improve modem emulation to get higher speeds.
  - Change default samplerates to 44100, blocksize to 1024 and prebuffer to 20,
    so that hopefully certain soundcards produce more fluent sound playback.
  - Add some rarely used, but for some games critical flags to 
    the internal commands.
  - Add -userconf flag, so that the userspecific configuration can
    easily be used together with -conf configfile.
  - Improve internal timing with repeating timers (especially with 
    the dynamic core).

0.73
  - Add two new opl2+opl3 emulators. (better speed, different implementation
    approach)
  - Improved DRO recording/better file structure.
  - Add EGA emulation.
  - Add special vga machine mode. Supports more of the exotic tricks like
    changing the palette during screen updates, 9x16 fonts etc.
  - Added special machine modes for the following svga cards:
     - S3
     - Paradise
     - Tseng
  - Fix problems with the vga split line feature.
  - Improve vesa emulation.
  - Add optional selection of old vesa mode for games that don't work
    with certain vesa features.

include/cpu.h
/*
 *  Copyright (C) 2002-2010  The DOSBox Team
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

/* $Id: cpu.h,v 1.57 2009-05-27 09:15:40 qbix79 Exp $ */

#ifndef DOSBOX_CPU_H
#define DOSBOX_CPU_H

#ifndef DOSBOX_DOSBOX_H
#include "dosbox.h" 
#endif
#ifndef DOSBOX_REGS_H
#include "regs.h"
#endif
#ifndef DOSBOX_MEM_H
#include "mem.h"
#endif

#define CPU_AUTODETERMINE_NONE		0x00
#define CPU_AUTODETERMINE_CORE		0x01
#define CPU_AUTODETERMINE_CYCLES	0x02

#define CPU_AUTODETERMINE_SHIFT		0x02
#define CPU_AUTODETERMINE_MASK		0x03

#define CPU_CYCLES_LOWER_LIMIT		200


#define CPU_ARCHTYPE_MIXED			0xff
#define CPU_ARCHTYPE_386SLOW		0x30
#define CPU_ARCHTYPE_386FAST		0x35
#define CPU_ARCHTYPE_486OLDSLOW		0x40
#define CPU_ARCHTYPE_486NEWSLOW		0x45
#define CPU_ARCHTYPE_PENTIUMSLOW	0x50

/* CPU Cycle Timing */
extern Bit32s CPU_Cycles;
extern Bit32s CPU_CycleLeft;
extern Bit32s CPU_CycleMax;
extern Bit32s CPU_OldCycleMax;
extern Bit32s CPU_CyclePercUsed;
extern Bit32s CPU_CycleLimit;
extern Bit64s CPU_IODelayRemoved;
extern bool CPU_CycleAutoAdjust;
extern bool CPU_SkipCycleAutoAdjust;
extern Bitu CPU_AutoDetermineMode;

extern Bitu CPU_ArchitectureType;

extern Bitu CPU_PrefetchQueueSize;

/* Some common Defines */
/* A CPU Handler */
typedef Bits (CPU_Decoder)(void);
extern CPU_Decoder * cpudecoder;

Bits CPU_Core_Normal_Run(void);
Bits CPU_Core_Normal_Trap_Run(void);
Bits CPU_Core_Simple_Run(void);
Bits CPU_Core_Full_Run(void);
Bits CPU_Core_Dyn_X86_Run(void);
Bits CPU_Core_Dyn_X86_Trap_Run(void);
Bits CPU_Core_Dynrec_Run(void);
Bits CPU_Core_Dynrec_Trap_Run(void);
Bits CPU_Core_Prefetch_Run(void);
Bits CPU_Core_Prefetch_Trap_Run(void);

void CPU_Enable_SkipAutoAdjust(void);
void CPU_Disable_SkipAutoAdjust(void);
void CPU_Reset_AutoAdjust(void);


//CPU Stuff

extern Bit16u parity_lookup[256];

bool CPU_LLDT(Bitu selector);
bool CPU_LTR(Bitu selector);
void CPU_LIDT(Bitu limit,Bitu base);
void CPU_LGDT(Bitu limit,Bitu base);

Bitu CPU_STR(void);
Bitu CPU_SLDT(void);
Bitu CPU_SIDT_base(void);
Bitu CPU_SIDT_limit(void);
Bitu CPU_SGDT_base(void);
Bitu CPU_SGDT_limit(void);

void CPU_ARPL(Bitu & dest_sel,Bitu src_sel);
void CPU_LAR(Bitu selector,Bitu & ar);
void CPU_LSL(Bitu selector,Bitu & limit);

void CPU_SET_CRX(Bitu cr,Bitu value);
bool CPU_WRITE_CRX(Bitu cr,Bitu value);
Bitu CPU_GET_CRX(Bitu cr);
bool CPU_READ_CRX(Bitu cr,Bit32u & retvalue);

bool CPU_WRITE_DRX(Bitu dr,Bitu value);
bool CPU_READ_DRX(Bitu dr,Bit32u & retvalue);

bool CPU_WRITE_TRX(Bitu dr,Bitu value);
bool CPU_READ_TRX(Bitu dr,Bit32u & retvalue);

Bitu CPU_SMSW(void);
bool CPU_LMSW(Bitu word);

void CPU_VERR(Bitu selector);
void CPU_VERW(Bitu selector);

void CPU_JMP(bool use32,Bitu selector,Bitu offset,Bitu oldeip);
void CPU_CALL(bool use32,Bitu selector,Bitu offset,Bitu oldeip);
void CPU_RET(bool use32,Bitu bytes,Bitu oldeip);
void CPU_IRET(bool use32,Bitu oldeip);
void CPU_HLT(Bitu oldeip);

bool CPU_POPF(Bitu use32);
bool CPU_PUSHF(Bitu use32);
bool CPU_CLI(void);
bool CPU_STI(void);

bool CPU_IO_Exception(Bitu port,Bitu size);
void CPU_RunException(void);

void CPU_ENTER(bool use32,Bitu bytes,Bitu level);

#define CPU_INT_SOFTWARE		0x1
#define CPU_INT_EXCEPTION		0x2
#define CPU_INT_HAS_ERROR		0x4
#define CPU_INT_NOIOPLCHECK		0x8

void CPU_Interrupt(Bitu num,Bitu type,Bitu oldeip);
static INLINE void CPU_HW_Interrupt(Bitu num) {
	CPU_Interrupt(num,0,reg_eip);
}
static INLINE void CPU_SW_Interrupt(Bitu num,Bitu oldeip) {
	CPU_Interrupt(num,CPU_INT_SOFTWARE,oldeip);
}
static INLINE void CPU_SW_Interrupt_NoIOPLCheck(Bitu num,Bitu oldeip) {
	CPU_Interrupt(num,CPU_INT_SOFTWARE|CPU_INT_NOIOPLCHECK,oldeip);
}

bool CPU_PrepareException(Bitu which,Bitu error);
void CPU_Exception(Bitu which,Bitu error=0);

bool CPU_SetSegGeneral(SegNames seg,Bitu value);
bool CPU_PopSeg(SegNames seg,bool use32);

bool CPU_CPUID(void);
Bitu CPU_Pop16(void);
Bitu CPU_Pop32(void);
void CPU_Push16(Bitu value);
void CPU_Push32(Bitu value);

void CPU_SetFlags(Bitu word,Bitu mask);


#define EXCEPTION_UD			6
#define EXCEPTION_TS			10
#define EXCEPTION_NP			11
#define EXCEPTION_SS			12
#define EXCEPTION_GP			13
#define EXCEPTION_PF			14

#define CR0_PROTECTION			0x00000001
#define CR0_MONITORPROCESSOR	0x00000002
#define CR0_FPUEMULATION		0x00000004
#define CR0_TASKSWITCH			0x00000008
#define CR0_FPUPRESENT			0x00000010
#define CR0_PAGING				0x80000000


// *********************************************************************
// Descriptor
// *********************************************************************

#define DESC_INVALID				0x00
#define DESC_286_TSS_A				0x01
#define DESC_LDT					0x02
#define DESC_286_TSS_B				0x03
#define DESC_286_CALL_GATE			0x04
#define DESC_TASK_GATE				0x05
#define DESC_286_INT_GATE			0x06
#define DESC_286_TRAP_GATE			0x07

#define DESC_386_TSS_A				0x09
#define DESC_386_TSS_B				0x0b
#define DESC_386_CALL_GATE			0x0c
#define DESC_386_INT_GATE			0x0e
#define DESC_386_TRAP_GATE			0x0f

/* EU/ED Expand UP/DOWN RO/RW Read Only/Read Write NA/A Accessed */
#define DESC_DATA_EU_RO_NA			0x10
#define DESC_DATA_EU_RO_A			0x11
#define DESC_DATA_EU_RW_NA			0x12
#define DESC_DATA_EU_RW_A			0x13
#define DESC_DATA_ED_RO_NA			0x14
#define DESC_DATA_ED_RO_A			0x15
#define DESC_DATA_ED_RW_NA			0x16
#define DESC_DATA_ED_RW_A			0x17

/* N/R Readable  NC/C Confirming A/NA Accessed */
#define DESC_CODE_N_NC_A			0x18
#define DESC_CODE_N_NC_NA			0x19
#define DESC_CODE_R_NC_A			0x1a
#define DESC_CODE_R_NC_NA			0x1b
#define DESC_CODE_N_C_A				0x1c
#define DESC_CODE_N_C_NA			0x1d
#define DESC_CODE_R_C_A				0x1e
#define DESC_CODE_R_C_NA			0x1f

#ifdef _MSC_VER
#pragma pack (1)
#endif

struct S_Descriptor {
#ifdef WORDS_BIGENDIAN
	Bit32u base_0_15	:16;
	Bit32u limit_0_15	:16;
	Bit32u base_24_31	:8;
	Bit32u g			:1;
	Bit32u big			:1;
	Bit32u r			:1;
	Bit32u avl			:1;
	Bit32u limit_16_19	:4;
	Bit32u p			:1;
	Bit32u dpl			:2;
	Bit32u type			:5;
	Bit32u base_16_23	:8;
#else
	Bit32u limit_0_15	:16;
	Bit32u base_0_15	:16;
	Bit32u base_16_23	:8;
	Bit32u type			:5;
	Bit32u dpl			:2;
	Bit32u p			:1;
	Bit32u limit_16_19	:4;
	Bit32u avl			:1;
	Bit32u r			:1;
	Bit32u big			:1;
	Bit32u g			:1;
	Bit32u base_24_31	:8;
#endif
}GCC_ATTRIBUTE(packed);

struct G_Descriptor {
#ifdef WORDS_BIGENDIAN
	Bit32u selector:	16;
	Bit32u offset_0_15	:16;
	Bit32u offset_16_31	:16;
	Bit32u p			:1;
	Bit32u dpl			:2;
	Bit32u type			:5;
	Bit32u reserved		:3;
	Bit32u paramcount	:5;
#else
	Bit32u offset_0_15	:16;
	Bit32u selector		:16;
	Bit32u paramcount	:5;
	Bit32u reserved		:3;
	Bit32u type			:5;
	Bit32u dpl			:2;
	Bit32u p			:1;
	Bit32u offset_16_31	:16;
#endif
} GCC_ATTRIBUTE(packed);

struct TSS_16 {	
    Bit16u back;                 /* Back link to other task */
    Bit16u sp0;				     /* The CK stack pointer */
    Bit16u ss0;					 /* The CK stack selector */
	Bit16u sp1;                  /* The parent KL stack pointer */
    Bit16u ss1;                  /* The parent KL stack selector */
	Bit16u sp2;                  /* Unused */
    Bit16u ss2;                  /* Unused */
    Bit16u ip;                   /* The instruction pointer */
    Bit16u flags;                /* The flags */
    Bit16u ax, cx, dx, bx;       /* The general purpose registers */
    Bit16u sp, bp, si, di;       /* The special purpose registers */
    Bit16u es;                   /* The extra selector */
    Bit16u cs;                   /* The code selector */
    Bit16u ss;                   /* The application stack selector */
    Bit16u ds;                   /* The data selector */
    Bit16u ldt;                  /* The local descriptor table */
} GCC_ATTRIBUTE(packed);

struct TSS_32 {	
    Bit32u back;                /* Back link to other task */
	Bit32u esp0;		         /* The CK stack pointer */
    Bit32u ss0;					 /* The CK stack selector */
	Bit32u esp1;                 /* The parent KL stack pointer */
    Bit32u ss1;                  /* The parent KL stack selector */
	Bit32u esp2;                 /* Unused */
    Bit32u ss2;                  /* Unused */
	Bit32u cr3;                  /* The page directory pointer */
    Bit32u eip;                  /* The instruction pointer */
    Bit32u eflags;               /* The flags */
    Bit32u eax, ecx, edx, ebx;   /* The general purpose registers */
    Bit32u esp, ebp, esi, edi;   /* The special purpose registers */
    Bit32u es;                   /* The extra selector */
    Bit32u cs;                   /* The code selector */
    Bit32u ss;                   /* The application stack selector */
    Bit32u ds;                   /* The data selector */
    Bit32u fs;                   /* And another extra selector */
    Bit32u gs;                   /* ... and another one */
    Bit32u ldt;                  /* The local descriptor table */
} GCC_ATTRIBUTE(packed);

#ifdef _MSC_VER
#pragma pack()
#endif
class Descriptor
{
public:
	Descriptor() { saved.fill[0]=saved.fill[1]=0; }

	void Load(PhysPt address);
	void Save(PhysPt address);

	PhysPt GetBase (void) { 
		return (saved.seg.base_24_31<<24) | (saved.seg.base_16_23<<16) | saved.seg.base_0_15; 
	}
	Bitu GetLimit (void) {
		Bitu limit = (saved.seg.limit_16_19<<16) | saved.seg.limit_0_15;
		if (saved.seg.g)	return (limit<<12) | 0xFFF;
		return limit;
	}
	Bitu GetOffset(void) {
		return (saved.gate.offset_16_31 << 16) | saved.gate.offset_0_15;
	}
	Bitu GetSelector(void) {
		return saved.gate.selector;
	}
	Bitu Type(void) {
		return saved.seg.type;
	}
	Bitu Conforming(void) {
		return saved.seg.type & 8;
	}
	Bitu DPL(void) {
		return saved.seg.dpl;
	}
	Bitu Big(void) {
		return saved.seg.big;
	}
public:
	union {
		S_Descriptor seg;
		G_Descriptor gate;
		Bit32u fill[2];
	} saved;
};

class DescriptorTable {
public:
	PhysPt	GetBase			(void)			{ return table_base;	}
	Bitu	GetLimit		(void)			{ return table_limit;	}
	void	SetBase			(PhysPt _base)	{ table_base = _base;	}
	void	SetLimit		(Bitu _limit)	{ table_limit= _limit;	}

	bool GetDescriptor	(Bitu selector, Descriptor& desc) {
		selector&=~7;
		if (selector>=table_limit) return false;
		desc.Load(table_base+(selector));
		return true;
	}
protected:
	PhysPt table_base;
	Bitu table_limit;
};

class GDTDescriptorTable : public DescriptorTable {
public:
	bool GetDescriptor(Bitu selector, Descriptor& desc) {
		Bitu address=selector & ~7;
		if (selector & 4) {
			if (address>=ldt_limit) return false;
			desc.Load(ldt_base+address);
			return true;
		} else {
			if (address>=table_limit) return false;
			desc.Load(table_base+address);
			return true;
		}
	}
	bool SetDescriptor(Bitu selector, Descriptor& desc) {
		Bitu address=selector & ~7;
		if (selector & 4) {
			if (address>=ldt_limit) return false;
			desc.Save(ldt_base+address);
			return true;
		} else {
			if (address>=table_limit) return false;
			desc.Save(table_base+address);
			return true;
		}
	} 
	Bitu SLDT(void)	{
		return ldt_value;
	}
	bool LLDT(Bitu value)	{
		if ((value&0xfffc)==0) {
			ldt_value=0;
			ldt_base=0;
			ldt_limit=0;
			return true;
		}
		Descriptor desc;
		if (!GetDescriptor(value,desc)) return !CPU_PrepareException(EXCEPTION_GP,value);
		if (desc.Type()!=DESC_LDT) return !CPU_PrepareException(EXCEPTION_GP,value);
		if (!desc.saved.seg.p) return !CPU_PrepareException(EXCEPTION_NP,value);
		ldt_base=desc.GetBase();
		ldt_limit=desc.GetLimit();
		ldt_value=value;
		return true;
	}
private:
	PhysPt ldt_base;
	Bitu ldt_limit;
	Bitu ldt_value;
};

class TSS_Descriptor : public Descriptor {
public:
	Bitu IsBusy(void) {
		return saved.seg.type & 2;
	}
	Bitu Is386(void) {
		return saved.seg.type & 8;
	}
	void SetBusy(bool busy) {
		if (busy) saved.seg.type|=2;
		else saved.seg.type&=~2;
	}
};


struct CPUBlock {
	Bitu cpl;							/* Current Privilege */
	Bitu mpl;
	Bitu cr0;
	bool pmode;							/* Is Protected mode enabled */
	GDTDescriptorTable gdt;
	DescriptorTable idt;
	struct {
		Bitu mask,notmask;
		bool big;
	} stack;
	struct {
		bool big;
	} code;
	struct {
		Bitu cs,eip;
		CPU_Decoder * old_decoder;
	} hlt;
	struct {
		Bitu which,error;
	} exception;
	Bits direction;
	bool trap_skip;
	Bit32u drx[8];
	Bit32u trx[8];
};

extern CPUBlock cpu;

static INLINE void CPU_SetFlagsd(Bitu word) {
	Bitu mask=cpu.cpl ? FMASK_NORMAL : FMASK_ALL;
	CPU_SetFlags(word,mask);
}

static INLINE void CPU_SetFlagsw(Bitu word) {
	Bitu mask=(cpu.cpl ? FMASK_NORMAL : FMASK_ALL) & 0xffff;
	CPU_SetFlags(word,mask);
}


#endif
